README.md
# node-red-node-ui-etable
This node is based on the capabilities of the Tabulator library and allows 
the user to define complex - editable tables and constructs.

Credit must go to Kazuhito Yokoi for his inspiration. node-red-node-ui-table

This node differs from node-red-node-ui-table due to it being more native to
the Tabulator libraries, it allows very complex editable tables to be built, although this means that the configuration of the node now must follow the http://tabulator.info json format.

Eg

Options can be

{
    "movableColumns": true,
    "resizableColumns": true,
    "selectable": 5,
    "responsiveLayout": "collapse",
    "autoResize": true,
    "layout": "fitColumns",
    "pagination": "local",
    "height": "800px",
    "groupBy": "system",
    "groupStartOpen": true
}

Columns can be

[		{title:"Name", field:"name", editor:"input"},
		{title:"Task Progress", field:"progress", align:"left", formatter:"progress", editor:true},
		{title:"Gender", field:"gender", width:95, editor:"select", editorParams:{values:["male", "female"]}},
		{title:"Rating", field:"rating", formatter:"star", align:"center", width:100, editor:true},
		{title:"Color", field:"col", width:130, editor:"input"},
		{title:"Date Of Birth", field:"dob", width:130, sorter:"date", align:"center"},
		{title:"Driver", field:"car", width:90,  align:"center", formatter:"tickCross", sorter:"boolean", editor:true}
]

Pass input data in msg.payload. Note that options and columns can also be passed in via msg. 
msg = {
    payload: ...
    config: {
        options: ...
        columns: ...
    }
}

CallBacks handled are :

        If the cell has editor = true,  a message is sent with "callback : cellEdited" 
        If the cell has editor = false, a message is sent with "callback : cellClick"



changes.md

V 1.0.3 Bump version for npmjs update issue
v 1.0.2	fixed emits not allowing table updates to browser added emitOnlyNewValues: false on widget startup

v.1.0.1 DYNAMIC ETables are now possible.
	msg.config.options and msg.config.columns are now configurable at runtime, this means that it is now possble to change the displayed tables based on button
	clicks or editable fields....on the fly !!	

v 0.1.5 Pull request by wisniaklo to add cellEdited, so, a new parameter is now added to indicate the callback.
        If the cell has editor = true,  a message is sent with "callback : cellEdited" 
	If the cell has editor = false, a message is sent with "callback : cellClick"

v 0.1.4 Allow specifying editor "position" for "select". this should be specified as part of editorParams: editorParams.position
  values are "top"(default), "center" / "middle" or 'bottom'
  
      {
        "title": "Event Type",
        "field": "eventType",
        "editor": "select",
        "width": "5%",
        "editorParams": {
            "values": [
                "PLANNED",
                "UNPLANNED"
            ],
            "position": "top"
        }
    } 
  
v 0.1.3 Initial release


dev/works/node.js
/**
 * Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

var path = require('path');

module.exports = function (RED) {
    function checkConfig(node, conf) {
        if (!conf || !conf.hasOwnProperty('group')) {
            node.error(RED._('table.error.no-group'));
            return false;
        }
        else {
            return true;
        }
    }

    function HTML(config,dark) {
        console.log('PRE=',config);
        var configAsJson = JSON.stringify(config);
        console.log('POST=',configAsJson);
        var mid = (dark) ? "_midnight" : "";
        var html = String.raw`
                <link href='ui-table/css/tabulator`+mid+`.min.css' rel='stylesheet' type='text/css'>
                <script type='text/javascript' src='ui-table/js/tabulator.js'></script>
                <div id='ui_table-{{$id}}'></div>
                <input type='hidden' ng-init='init(` + configAsJson + `)'>
            `;
        return html;
    };

    function TableNode(config) {
        var done = null;
        var node = this;
        try {
            RED.nodes.createNode(this, config);
            if (checkConfig(node, config)) {
                var ui = RED.require('node-red-dashboard')(RED);
                console.log(config);
                var luma = 255;
                if (ui.hasOwnProperty("getTheme") && (ui.getTheme() !== undefined)) {
                    var rgb = parseInt(ui.getTheme()["page-sidebar-backgroundColor"].value.substring(1), 16);   // convert rrggbb to decimal
                    luma = 0.2126 * ((rgb >> 16) & 0xff) + 0.7152 * ((rgb >>  8) & 0xff) + 0.0722 * ((rgb >>  0) & 0xff); // per ITU-R BT.709
                }
                if (config.height == 0) { config.height = 20; } // min height to 2 so auto will show something

                config.columns = JSON.parse(config.payload);
                delete config.payload;
                var html = HTML(config,(luma < 128));
                console.log('Columns after HTML=',config);
                console.log('HTML=',html);
                console.log('Columns=',config);
                done = ui.addWidget({
                    node: node,
                    width: config.width,
                    height: config.height,
                    format: html,
                    templateScope: 'local',
                    order: config.order,
                    group: config.group,
                    forwardInputMessages: false,
                    beforeEmit: function (msg, value) {
                        return { msg: { payload: value } };
                    },
                    beforeSend: function (msg, orig) {
                        if (orig) { return orig.msg; }
                    },
                    initController: function ($scope, events) {
                        $scope.inited = false;
                        $scope.tabledata = [];
                        var tablediv;
                        var createTable = function(basediv, tabledata, columndata, outputs) {
                            var y = (columndata.length === 0) ? 25 : 32;
                            var opts = {
                                data: tabledata,
                                layout: 'fitColumns',
                                columns: columndata,
                                autoColumns: columndata.length == 0,
                                movableColumns: true,
addRowPos:"bottom",
pagination:"local",
paginationSize:20,
                               height: tabledata.length * y + 26
                            }
console.log(opts);
                            if (outputs > 0) {
                                opts.cellClick = function(e, cell) {
                                    $scope.send({topic:cell.getField(),payload:cell.getData()});
                                };
                            }
                            var table = new Tabulator(basediv, opts);
                            $("#add-row").click(function(){table.addRow({});});
                            $("#del-row").click(function(){table.deleteRow({});});
                            $("#clear-row").click(function(){table.clearRow({});});
                        };
                        $scope.init = function (config) {
                            $scope.config = config;
                            tablediv = '#ui_table-' + $scope.$eval('$id')
                            var stateCheck = setInterval(function() {
                                if (document.querySelector(tablediv) && $scope.tabledata) {
                                    clearInterval(stateCheck);
                                    $scope.inited = true;
                                    createTable(tablediv,$scope.tabledata,$scope.config.columns,$scope.config.outputs);
                                    $scope.tabledata = [];
                                }
                            }, 40);
                        };
                        $scope.$watch('msg', function (msg) {
                            if (msg && msg.hasOwnProperty("payload") && Array.isArray(msg.payload)) {
                                if ($scope.inited == false) {
                                    $scope.tabledata = msg.payload;
                                    return;
                                }
console.log('MSG=',msg.payload,'\n-',$scope.config.columns);
                                createTable(tablediv,msg.payload,$scope.config.columns,$scope.config.outputs);
                            }
                        });
                    }
                });
            }
        }
        catch (e) { console.log(e); }

        node.on('close', function () {
            if (done) { done(); }
        });
    }

    RED.nodes.registerType('ui_table', TableNode);

    var uipath = 'ui';
    if (RED.settings.ui) { uipath = RED.settings.ui.path; }
    var fullPath = path.join(RED.settings.httpNodeRoot, uipath, '/ui-table/*').replace(/\\/g, '/');;
    RED.httpNode.get(fullPath, function (req, res) {
        var options = {
            root: __dirname + '/lib/',
            dotfiles: 'deny'
        };
        res.sendFile(req.params[0], options)
    });
};


etable.js
/**
 * Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

var path = require('path');

module.exports = function (RED) {
    function checkConfig(node, conf) {
        if (!conf || !conf.hasOwnProperty('group')) {
            node.error(RED._('table.error.no-group'));
            return false;
        }
        else {
            return true;
        }
    }

    function HTML(config,dark) {
        var configAsJson = JSON.stringify(config);
        var mid = (dark) ? "_midnight" : "";
        var html = String.raw`
                <link href='ui-etable/css/tabulator`+mid+`.min.css' rel='stylesheet' type='text/css'>
                <script type='text/javascript' src='ui-etable/js/tabulator.js'></script>
                <div id='ui_etable-{{$id}}'></div>
                <input type='hidden' ng-init='init(` + configAsJson + `)'>
            `;
        return html;
    };

    function eTableNode(config) {
        var done = null;
        var node = this;
        try {
            RED.nodes.createNode(this, config);
            if (checkConfig(node, config)) {
                var ui = RED.require('node-red-dashboard')(RED);

                var luma = 255;
                if (ui.hasOwnProperty("getTheme") && (ui.getTheme() !== undefined)) {
                    var rgb = parseInt(ui.getTheme()["page-sidebar-backgroundColor"].value.substring(1), 16);   // convert rrggbb to decimal
                    luma = 0.2126 * ((rgb >> 16) & 0xff) + 0.7152 * ((rgb >>  8) & 0xff) + 0.0722 * ((rgb >>  0) & 0xff); // per ITU-R BT.709
                }
                if (config.height == 0) { config.height = 20; } // min height to 2 so auto will show something

                config.columns = JSON.parse(config.payload);
                delete config.payload;
                delete config.payloadType;
                config.options = JSON.parse(config.options);
                var html = HTML(config,(luma < 128));

                done = ui.addWidget({
                    node: node,
                    width: config.width,
                    height: config.height,
                    format: html,
                    templateScope: 'local',
                    order: config.order,
                    group: config.group,
                    forwardInputMessages: false,
                    emitOnlyNewValues: false,
                    beforeEmit: function (msg, value) {
                        return {
                            msg: {
                                payload: value,
                                config: msg.config
                            }
                        };
                    },
                    beforeSend: function (msg, orig) {
                        if (orig) { return orig.msg; }
                    },
                    initController: function ($scope, events) {
                        $scope.inited = false;
                        $scope.tabledata = [];
                        var tablediv;
                        var createTable = function(basediv, tabledata, columndata, options, outputs) {
                            var y = (columndata.length === 0) ? 25 : 32;
                            var opts1 = {
                                data: tabledata,
                                columns: columndata,
                                autoColumns: columndata.length == 0,
                                height: tabledata.length * y + 26
                            }
                            var opts = Object.assign(opts1,options);
                            if (outputs > 0) {
                                opts.cellClick  = function(e,cell) {
                                    $scope.send({topic:cell.getField(),callback:"cellClick",payload:cell.getData(),options:opts});
                                };
                                opts.cellEdited = function(cell) {
                                    $scope.send({topic:cell.getField(),callback:"cellEdited",payload:cell.getData(),options:opts});
                                };
                            }
                            var table = new Tabulator(basediv, opts);
                        };
                        $scope.init = function (config) {
                            $scope.config = config;
                            tablediv = '#ui_etable-' + $scope.$eval('$id')
                            var stateCheck = setInterval(function() {
                                if (document.querySelector(tablediv) && $scope.tabledata) {
                                    clearInterval(stateCheck);
                                    $scope.inited = true;
                                    createTable(tablediv,$scope.tabledata,$scope.config.columns,$scope.config.options,$scope.config.outputs);
                                    $scope.tabledata = [];
                                }
                            }, 40);
                        };
                        $scope.$watch('msg', function (msg) {
                            var columns = $scope.config.columns;
                            var options = $scope.config.options;
                            if(msg && msg.hasOwnProperty("config")){
                                if(msg.config.options){
                                    options = msg.config.options;
                                }
                                if(msg.config.columns){
                                    columns = msg.config.columns;
                                }
                            }
                            if (msg && msg.hasOwnProperty("payload") && Array.isArray(msg.payload)) {
                                if ($scope.inited == false) {
                                    $scope.tabledata = msg.payload;
                                    return;
                                }
                                createTable(tablediv,msg.payload,columns,options,$scope.config.outputs);
                            }
                        });
                    }
                });
            }
        }
        catch (e) { console.log(e); }

        node.on('close', function () {
            if (done) { done(); }
        });
    }

    RED.nodes.registerType('ui_etable', eTableNode);

    var uipath = 'ui';
    if (RED.settings.ui) { uipath = RED.settings.ui.path; }
    var fullPath = path.join(RED.settings.httpNodeRoot, uipath, '/ui-etable/*').replace(/\\/g, '/');;
    RED.httpNode.get(fullPath, function (req, res) {
        var options = {
            root: __dirname + '/lib/',
            dotfiles: 'deny'
        };
        res.sendFile(req.params[0], options)
    });
};
